### PR TITLE
BUILD: Make autogen.sh pull git submodules

### PR DESCRIPTION
**Pull request by MattBBaker**
Updates the autogen.sh to check for the existence of git and do git submodule commands to pull all of the defined submodules.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1077/ for details.

File: autogen.sh [hunk 1]
```diff
@@ -3,6 +3,15 @@
 rm -rf autom4te.cache
 mkdir -p config/m4 config/aux external/jemalloc
 
+GIT_FOUND=1
+command git --help 2>&1 > /dev/null || { echo "Git not found, not retrieving external packages"; export GIT_FOUND=0; }
```
Comment (shamisp) [inline autogen.sh]: 1. Please avoid long lines
2. You don't really have to export since this is an internal variable
3. Please add more descriptive error saying that you not going to install jemalloc

**Comment by shamisp** [comment]
autobot build

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/508/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
```
install -m 644 doc/jemalloc.html /ccs/home/buildbot/buildbot/pod7/ucx-rhel5-x86_64/build/../install/share/doc/jemalloc
install: cannot stat `doc/jemalloc.html': No such file or directory
make[2]: *** [install_doc_html] Error 1
make[2]: Leaving directory `/autofs/nccs-svm1_home1/buildbot/buildbot/pod7/ucx-rhel5-x86_64/build/external/jemalloc'
make[1]: *** [install-recursive] Error 1
make[1]: Leaving directory `/autofs/nccs-svm1_home1/buildbot/buildbot/pod7/ucx-rhel5-x86_64/build/external'
make: *** [install-recursive] Error 1
program finished with exit code 2
elapsedTime=9.181243
```

**Comment by MattBBaker** [comment]
I was just looking into that. It looks like jemalloc doesn't actually build some html doc files that it needs for install for reasons that I don't understand.

**Comment by MattBBaker** [comment]
This looks like a bug in jemalloc. I'm poking them about it. I have a work around, but I'm going to wait for a bit before committing it to see what the jemalloc people say.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1085/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/515/ for details (Mellanox internal link).

**Comment by MattBBaker** [comment]
@shamisp Updated with your concerns. It looks like the mellanox failures come from coverity scans on jemalloc. Is there a way to suppress this?

**Comment by shamisp** [comment]
@yosefe ?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1086/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/516/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@miked-mellanox can you please take a look?

**Comment by yosefe** [comment]
@MattBBaker pls update scripts to exclude jemalloc from coverity

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1089/ for details.

**Comment by shamisp** [comment]
1. I like the [-d .git]
2. With submodule we explicitly specify the version that we want to use (4.0.4). 

IMHO it is much cleaner that taball. Also this is not an external dependency but an optimization + library we consider to re-distribute. Unless we modify the library I don't see a reason why we have to keep the tarball in our code base.

**Comment by MattBBaker** [comment]
1. I agree on the -d .git
2. The git submodule will not track HEAD, rather it tracks a specific commit. In this case, the 4.0.4 release. 

Question, is the test_jenkins.sh script what actually gets run? If so I may have a solution.

**Comment by shamisp** [comment]
@MattBBaker Yes, jenkins actually runs the script.

**Comment by shamisp** [comment]
@yosefe and @MattBBaker  - please agree on the coverity issue.

**Comment by shamisp** [comment]
@MattBBaker @miked-mellanox @yosefe  - what is the status here ?

**Comment by miked-mellanox** [comment]
IMO, -d .git should be checked in addition that git.exe found

**Comment by MattBBaker** [comment]
I have an update that will do this, but wanted to wait until we could figure out a way to make coverty ignore jemalloc before committing it.

**Comment by shamisp** [comment]
Who is handling coverty  ignore ?

**Comment by MattBBaker** [comment]
I'm not sure. I can't work on it because I don't have access to anything with coverty and I've not heard back about anyone poking it.

**Comment by miked-mellanox** [comment]
i`ll take a look on coveity issues tomorrow.

**Comment by shamisp** [comment]
@miked-mellanox - this one is waiting for you :smiley:

**Comment by miked-mellanox** [comment]
@shamisp - https://github.com/openucx/ucx/pull/503 should be merged to support coverity exclude.

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1136/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/553/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
it should pass now, right ?

**Comment by MattBBaker** [comment]
Still needed the check for .git. It's there now.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1137/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/554/ for details (Mellanox internal link).

**Comment by miked-mellanox** [comment]
failed on ubunty

``` shell
7:06:02 Autogen
07:06:02 + ./autogen.sh
07:06:02 Submodule 'external/jemalloc' (git@github.com:jemalloc/jemalloc.git) registered for path 'external/jemalloc'
07:06:02 Cloning into 'external/jemalloc'...
07:06:09 Submodule path 'external/jemalloc': checked out '91010a9e2ebfc84b1ac1ed7fdde3bfed4f65f180'
07:06:09 ./autogen.sh: 19: ./autogen.sh: pushd: not found
07:06:09 configure.ac:37: error: possibly undefined macro: AM_INIT_AUTOMAKE
07:06:09       If this token and others are legitimate, please use m4_pattern_allow.
07:06:09       See the Autoconf documentation.
07:06:09 configure.ac:39: error: possibly undefined macro: AM_MAINTAINER_MODE
07:06:09 configure.ac:53: error: possibly undefined macro: AC_SUBST
07:06:09 configure.ac:63: error: possibly undefined macro: AM_PROG_AS
07:06:09 configure.ac:66: error: possibly undefined macro: AC_PROG_LIBTOOL
07:06:09 configure.ac:115: error: possibly undefined macro: AS_IF
07:06:09 configure.ac:133: error: possibly undefined macro: AC_DEFINE
07:06:09 configure.ac:188: error: possibly undefined macro: AM_CONDITIONAL
07:06:09 configure:6836: error: possibly undefined macro: AC_MSG_WARN
07:06:09 configure:6837: error: possibly undefined macro: AC_MSG_ERROR
07:06:09 Build step 'Execute shell' marked build as failure
07:06:09 TAP Reports Processing: START
07:06:09 Looking for TAP results report in workspace using pattern: **/*.tap
07:06:09 Did not find any matching files.
```

**Comment by shamisp** [comment]
Just replace sh to #!/bin/bash.
pushd is internal command

**Comment by shamisp** [comment]
AFAIK there is no replacement for pushd , unless you cache the olddir in variable , change to new one and than go back to old

**Comment by miked-mellanox** [comment]
yep, bash may work

**Comment by shamisp** [comment]
@MattBBaker - can u please fix ?

**Comment by MattBBaker** [comment]
Done.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1138/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/555/ for details (Mellanox internal link).

**Comment by miked-mellanox** [comment]
failure unrelated to this PR

``` shell
20:06:04 [ RUN      ] cm/test_many2one_am.am_bcopy/0
20:06:05 /scrap/jenkins/scrap/workspace/hpc-ucx-pr/label/hpc-test-node/contrib/../src/ucs/gtest/test.cc:111: Failure
20:06:05 Failed
20:06:05 Got 1 warnings during the test
20:06:05 
[  FAILED  ] cm/test_many2one_am.am_bcopy/0, where GetParam() = cm/mlx5_0:1 (1542 ms)
20:06:05 [ RUN      ] cm/test_many2one_am.am_bcopy/1
```

File: autogen.sh [hunk 2]
```diff
@@ -1,8 +1,19 @@
...
 rm -rf autom4te.cache
 mkdir -p config/m4 config/aux external/jemalloc
 
+GIT_FOUND=1
+command git --help 2>&1 > /dev/null || { \
+    echo "Git not found, will not retrieve jemalloc"; \
+    GIT_FOUND=0; }
+
+if [ $GIT_FOUND -eq 1 -a -d .git ]
+then
```
Comment (yosefe) [inline autogen.sh]: indentation

File: autogen.sh [hunk 3]
```diff
@@ -1,8 +1,19 @@
-#!/bin/sh
+#!/bin/bash
 
 rm -rf autom4te.cache
 mkdir -p config/m4 config/aux external/jemalloc
 
+GIT_FOUND=1
+command git --help 2>&1 > /dev/null || { \
```
Comment (yosefe) [inline autogen.sh]: can use >&/dev/null

File: autogen.sh [hunk 4]
```diff
@@ -1,8 +1,19 @@
...
+    echo "Git not found, will not retrieve jemalloc"; \
```
Comment (yosefe) [inline autogen.sh]: maybe better to say "will not retrieve external sub-projects"

File: autogen.sh [hunk 5]
```diff
@@ -1,8 +1,19 @@
...
+    GIT_FOUND=0; }
+
+if [ $GIT_FOUND -eq 1 -a -d .git ]
+then
+git submodule init
+git submodule update
```
Comment (yosefe) [inline autogen.sh]: do we need "git submodule sync"?

File: autogen.sh [hunk 6]
```diff
@@ -1,8 +1,19 @@
...
+fi
+
 if [ -e external/jemalloc/configure.ac ]
 then
```
Comment (yosefe) [inline autogen.sh]: indentation

File: external/Makefile.am [hunk 7]
```diff
@@ -7,3 +13,15 @@ endif
...
 # to do distcheck or distirbute anything in external
 DIST_SUBDIRS=
+
+# If you want to just skip testing, but still package the 
+# subdir, add it here
+dist-hook:
+	mkdir -p $(distdir)/jemalloc/
+	cp -r $(srcdir)/jemalloc/* $(distdir)/jemalloc/
+
+# Custom rules for installing external libs.
```
Comment (yosefe) [inline external/Makefile.am]: comment seems irrelevant. copy-paste?

File: external/Makefile.am [hunk 8]
```diff
@@ -7,3 +13,15 @@ endif
 # This is intentionally left empty. This will dev/null any attempt
 # to do distcheck or distirbute anything in external
 DIST_SUBDIRS=
+
+# If you want to just skip testing, but still package the 
+# subdir, add it here
```
Comment (yosefe) [inline external/Makefile.am]: comment seems irrelevant. copy-paste?

File: autogen.sh (hunk 3 above)
Comment (shamisp) [inline autogen.sh]: This is pretty much "religious" argument. I actually prefer the "textbook" syntax `2>&1`

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/556/ for details (Mellanox internal link).

File: external/Makefile.am (hunk 8 above)
Comment (MattBBaker) [inline external/Makefile.am]: I would argue this one is needed. The dist-hook target is pretty generic, so explaining what this one is for I think is worthwhile.

File: external/Makefile.am (hunk 7 above)
Comment (MattBBaker) [inline external/Makefile.am]: This one can go, sure.

**Comment by shamisp** [comment]
@yosefe - can we merge it in ?

**Comment by yosefe** [comment]
@MattBBaker pls fix minor comments

File: autogen.sh (hunk 5 above)
Comment (MattBBaker) [inline autogen.sh]: I looked into it, and it looks like sync is only needed if you are changing the URL in the submodule file for a project that was already initialized. It is not needed on the first initialization.

**Comment by MattBBaker** [comment]
@shamisp @yosefe Fixed up the comments.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1144/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/562/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
@yosefe - ready to go ?

**Comment by yosefe** [comment]
:+1:

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1147/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1157/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1172/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1173/ for details.
